File: April 2024 GFG SOLUTION/April-17.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG
{
	public static void main(String[] args) throws IOException
	{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while(t-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            int a[] = new int[(int)n];
            String inputLine[] = br.readLine().trim().split(" ");
            for (int i = 0; i < n; i++) {
                a[i] = Integer.parseInt(inputLine[i]);
            }
            
            Solution obj = new Solution();  
            System.out.println(obj.countPairs(a, n));
            
        }
	}
}

// } Driver Code Ends


//User function Template for Java

class Solution {
    static int merge(int arr[], int temp[], int left, int mid, int right) {
        int inv_count = 0;
        int i = left;
        int j = mid;
        int k = left;

        while ((i <= mid - 1) && (j <= right)) {
            if (arr[i] <= arr[j])
                temp[k++] = arr[i++];
            else {
                temp[k++] = arr[j++];
                inv_count = inv_count + (mid - i);
            }
        }
        while (i <= mid - 1)
            temp[k++] = arr[i++];
        while (j <= right)
            temp[k++] = arr[j++];

        for (i = left; i <= right; i++)
            arr[i] = temp[i];

        return inv_count;
    }

    static int _mergeSort(int arr[], int temp[], int left, int right) {
        int mid, inv_count = 0;
        if (right > left) {
            mid = (right + left) / 2;

            inv_count = _mergeSort(arr, temp, left, mid);
            inv_count += _mergeSort(arr, temp, mid + 1, right);

            inv_count += merge(arr, temp, left, mid + 1, right);
        }

        return inv_count;
    }

    static int countPairs(int arr[], int n) {
        for (int i = 0; i < n; i++)
            arr[i] = i * arr[i];
        int temp[] = new int[n];
        return _mergeSort(arr, temp, 0, n - 1);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            int n = sc.nextInt();
            int a[] = new int[n];
            for (int i = 0; i < n; i++) {
                a[i] = sc.nextInt();
            }
            System.out.println(countPairs(a, n));
        }
    }
}

File: April 2024 GFG SOLUTION/April-17.py
#User function Template for python3

class Solution:
    def merge(self, arr, temp, left, mid, right):
        inv_count = 0
        i = left
        j = mid
        k = left
        while i <= mid - 1 and j <= right:
            if arr[i] <= arr[j]:
                temp[k] = arr[i]
                i += 1
                k += 1
            else:
                temp[k] = arr[j]
                k += 1
                j += 1
                inv_count += mid - i
        while i <= mid - 1:
            temp[k] = arr[i]
            i += 1
            k += 1
        while j <= right:
            temp[k] = arr[j]
            k += 1
            j += 1
        for i in range(left, right + 1):
            arr[i] = temp[i]
        return inv_count

    def _mergeSort(self, arr, temp, left, right):
        inv_count = 0
        if right > left:
            mid = (right + left) // 2
            inv_count = self._mergeSort(arr, temp, left, mid)
            inv_count += self._mergeSort(arr, temp, mid + 1, right)
            inv_count += self.merge(arr, temp, left, mid + 1, right)
        return inv_count

    def countPairs(self, arr, n):
        for i in range(n):
            arr[i] = i * arr[i]
        temp = [0] * n
        return self._mergeSort(arr, temp, 0, n - 1)


#{ 
 # Driver Code Starts
#Initial Template for Python 3

def main():

    T = int(input())

    while(T > 0):
        n = int(input())
        a = [int(x) for x in input().strip().split()]
        ob= Solution()
        print(ob.countPairs(a, n))

        T -= 1


if __name__ == "__main__":
    main()
    
# } Driver Code Ends
